File: Алгоритмизация/Практики/Практическая №7/4/main.py
import sys

def is_safe(v: int, graph: list, color: list, c: int, n: int) -> bool:
    for i in range(n):
        if graph[v][i] and c == color[i]:
            return False
    return True

def graph_coloring_util(graph: list, m: int, color: list, v: int, n: int) -> bool:
    if v == n:
        return True
    for c in range(1, m + 1):
        if is_safe(v, graph, color, c, n):
            color[v] = c
            if graph_coloring_util(graph, m, color, v + 1, n):
                return True
            color[v] = 0
    return False

def graph_coloring(graph: list, m: int, color: list, n: int) -> bool:
    color[:] = [0] * n
    return graph_coloring_util(graph, m, color, 0, n)

if __name__ == "__main__":
    input_data = sys.stdin.read().split()
    if not input_data:
        sys.exit(0)
    n = int(input_data[0])
    k = int(input_data[1])
    graph = []
    idx = 2
    for i in range(n):
        row_str = input_data[idx]
        graph.append([int(char) for char in row_str])
        idx += 1
    color = [0] * n
    if graph_coloring(graph, k, color, n):
        print("YES")
        print(" ".join(map(str, color)))
    else:
        print("NO")
      

File: Алгоритмизация/Практики/Практическая №7/4/main.cpp
#include <iostream>
#include <vector>
#include <string>

bool isSafe(int v, const std::vector<std::vector<int>>& graph, const std::vector<int>& color, int c, int n) {
    for (int i = 0; i < n; ++i) {
        if (graph[v][i] && c == color[i]) {
            return false;
        }
    }
    return true;
}

bool graphColoringUtil(const std::vector<std::vector<int>>& graph, int m, std::vector<int>& color, int v, int n) {
    if (v == n) {
        return true;
    }
    for (int c = 1; c <= m; ++c) {
        if (isSafe(v, graph, color, c, n)) {
            color[v] = c;
            if (graphColoringUtil(graph, m, color, v + 1, n)) {
                return true;
            }
            color[v] = 0;
        }
    }
    return false;
}

bool graphColoring(const std::vector<std::vector<int>>& graph, int m, std::vector<int>& color, int n) {
    color.assign(n, 0);
    return graphColoringUtil(graph, m, color, 0, n);
}

int main() {
    int n, k;
    if (!(std::cin >> n >> k)) return 0;
    std::vector<std::vector<int>> graph(n, std::vector<int>(n));
    for (int i = 0; i < n; ++i) {
        std::string line;
        std::cin >> line;
        for (int j = 0; j < n; ++j) {
            graph[i][j] = line[j] - '0';
        }
    }
    std::vector<int> color(n, 0);
    if (graphColoring(graph, k, color, n)) {
        std::cout << "YES" << std::endl;
        for (int i = 0; i < n; ++i) {
            std::cout << color[i] << (i == n - 1 ? "" : " ");
        }
        std::cout << std::endl;
    }
    else {
        std::cout << "NO" << std::endl;
    }
    return 0;
}
